statistics/statistics.go

```diff
@@ -21,329 +21,143 @@ package statistics
-	min, minErr := stats.Min(buffer)
-	max, maxErr := stats.Max(buffer)
+}
 
-	for _, stat := range statList {
-		switch stat {
-		case "count":
-			val = float64(len(buffer))
-		case "mean":
-			val, err = stats.Mean(buffer)
-		case "median":
-			val, err = stats.Median(buffer)
-		case "standard_deviation":
-			val, err = stats.StandardDeviation(buffer)
-		case "variance":
-			val, err = stats.Variance(buffer)
-		case "95%_ile":
-			val, err = stats.PercentileNearestRank(buffer, 95)
-		case "99%_ile":
-			val, err = stats.PercentileNearestRank(buffer, 99)
-		case "2%_ile":
-			val, err = stats.PercentileNearestRank(buffer, 2)
-		case "9%_ile":
-			val, err = stats.PercentileNearestRank(buffer, 9)
-		case "25%_ile":
-			val, err = stats.PercentileNearestRank(buffer, 25)
-		case "75%_ile":
-			val, err = stats.PercentileNearestRank(buffer, 75)
-		case "91%_ile":
-			val, err = stats.PercentileNearestRank(buffer, 91)
-		case "98%_ile":
-			val, err = stats.PercentileNearestRank(buffer, 98)
-		case "minimum":
-			val, err = min, minErr
-		case "maximum":
-			val, err = max, maxErr
-		case "range":
-			val = max - min
-		case "mode":
-			modeVal, err = stats.Mode(buffer)
-		case "kurtosis":
-			val, err = p.Kurtosis(buffer)
-		case "skewness":
-			val, err = p.Skewness(buffer)
-		case "sum":
-			val, err = stats.Sum(buffer)
-		case "trimean":
-			val, err = stats.Trimean(buffer)
-		case "quartile_range":
-			val, err = stats.InterQuartileRange(buffer)
-		default:
-			st := fmt.Sprintf("Unknown statistic received %T:", stat)
-			log.Errorf(st)
-			err = errors.New(st)
+// Process processes the data, inputs the data into sorted buffer and calls the GetStats method
+func (p *Plugin) Process(metrics []plugin.Metric, cfg plugin.Config) ([]plugin.Metric, error) {
+	var result []plugin.Metric
+	for _, metric := range metrics {
+		slidingWindowLength, slidingFactor, stats, err := GetConfig(metric.Config)
+		if err != nil {
+			return nil, err
 		}
 
+		// convert any number to float64
+		floatValue, err := dataToFloat64(metric.Data)
 		if err != nil {
-			log.Warnf("Error in %T", stat)
+			return nil, err
 		}
 
-		metric := plugin.Metric{
-			Data:      val,
-			Namespace: plugin.NewNamespace(namespace, stat),
-			Timestamp: time,
-			Unit:      unit,
-			Tags:      tags,
-		}
+		nsSlice := metric.Namespace.Strings()
+		ns := strings.Join(nsSlice, "")
+		_, ok := p.buffer[ns]
+		if !ok {
+			//buffer doesn't exist, so create one
+			p.buffer[ns] = &dataBuffer{
+				data:            make([]*data, 0, slidingWindowLength),
+				dataByTimestamp: make([]*data, 0, slidingWindowLength)}
 
-		if stat == "mode" {
-			metric.Data = modeVal
+		} else {
+			if slidingWindowLength != cap(p.buffer[ns].data) {
```

Comment:
This is really a futuristic feature where we could dynamically update the slidingWindowLength config (thing we can do from the REST API). In that case we need to resize the buffer to the new lenght.

Currently those condition are removed by SSA when we compile, the idea is to implement this feature after refactoring the plugin.